File: javascript/easy/averageOfLevelsInBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function(root) {
    const queue = new Array(), res = new Array();
    if(root !== null) {
        queue.push(root);
    }
    let currentCount = 0, sum = 0, numNodesInLevel = queue.length;
    while(queue.length > 0) {
        const current = queue.shift();
        sum += current.val;
        currentCount++;
        if(current.left !== null) {
            queue.push(current.left);
        }
        if(current.right !== null) {
            queue.push(current.right);
        }
        if(currentCount === numNodesInLevel) {
            res.push(sum/numNodesInLevel);
            sum = currentCount = 0;
            numNodesInLevel = queue.length;
        }
    }
    return res;
};

File: ruby/easy/average_of_levels_in_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Float[]}
def average_of_levels(root)
    res = Array.new
    queue = Array.new
    queue.push(root) if root
    i, sum, num_nodes_in_level = 0, 0, queue.length
    while !queue.empty?
        current = queue.shift
        sum += current.val
        i += 1
        queue.push(current.left) if current.left
        queue.push(current.right) if current.right
        if i == num_nodes_in_level
            res.push((sum.to_f/num_nodes_in_level))
            num_nodes_in_level = queue.length
            i = sum = 0
        end
    end
    return res
end
